packages/lodestar/src/chain/prepareNextSlot.ts

```diff
@@ -0,0 +1,129 @@
+import {computeEpochAtSlot, isBellatrixStateType} from "@chainsafe/lodestar-beacon-state-transition";
+import {IChainForkConfig} from "@chainsafe/lodestar-config";
+import {ForkSeq, SLOTS_PER_EPOCH} from "@chainsafe/lodestar-params";
+import {Slot} from "@chainsafe/lodestar-types";
+import {ILogger, sleep} from "@chainsafe/lodestar-utils";
+import {GENESIS_EPOCH, ZERO_HASH_HEX} from "../constants/constants.js";
+import {IMetrics} from "../metrics/index.js";
+import {ChainEvent} from "./emitter.js";
+import {prepareExecutionPayload} from "./factory/block/body.js";
+import {IBeaconChain} from "./interface.js";
+import {RegenCaller} from "./regen/index.js";
+
+/* With 12s slot times, this scheduler will run 4s before the start of each slot (`12 / 3 = 4`). */
+const SCHEDULER_LOOKAHEAD_FACTOR = 3;
+
+/* We don't want to do more epoch transition than this */
+const PREPARE_EPOCH_LIMIT = 1;
+
+/**
+ * At Bellatrix, if we are responsible for proposing in next slot, we want to prepare payload
+ * 4s (1/3 slot) before the start of next slot
+ *
+ * For all forks, when clock is 1/3 slot before an epoch, we want to prepare for the next epoch
+ * transition from our head so that:
+ * + validators vote for block head on time through attestation
+ * + validators propose blocks on time
+ * + For Bellatrix, to compute proposers of next epoch so that we can prepare new payloads
+ *
+ */
+export class PrepareNextSlotScheduler {
+  constructor(
+    private readonly chain: IBeaconChain,
+    private readonly config: IChainForkConfig,
+    private readonly metrics: IMetrics | null,
+    private readonly logger: ILogger,
+    private readonly signal: AbortSignal
+  ) {
+    this.chain.emitter.on(ChainEvent.clockSlot, this.prepareForNextSlot);
+
+    this.signal.addEventListener(
+      "abort",
+      () => {
+        this.chain.emitter.off(ChainEvent.clockSlot, this.prepareForNextSlot);
+      },
+      {once: true}
+    );
+  }
+
+  /**
+   * Use clockSlot instead of clockEpoch to schedule the task at more exact time.
+   */
+  prepareForNextSlot = async (clockSlot: Slot): Promise<void> => {
+    const prepareSlot = clockSlot + 1;
+    const prepareEpoch = computeEpochAtSlot(prepareSlot);
+    const isLastEpochSlot = (clockSlot + 1) % SLOTS_PER_EPOCH === 0;
+    if (this.config.getForkSeq(prepareEpoch) < ForkSeq.bellatrix && !isLastEpochSlot) {
+      return;
+    }
+
+    const slotMs = this.config.SECONDS_PER_SLOT * 1000;
+    // At 1/3 slot time before the next slot, we either prepare payload or precompute epoch transition
+    await sleep(slotMs - slotMs / SCHEDULER_LOOKAHEAD_FACTOR, this.signal);
+
+    const {slot: headSlot, blockRoot: headRoot} = this.chain.forkChoice.getHead();
```

Comment:
should we try optimizing head for next slot, i.e. process attestations of this `clockSlot` as well and call updateHead. it would reduce any reorgs possibility immensely.